Support multiple registry auth via environment variables

DennisDenuto opened the pull request:
This is a draft PR to increase visibility on changes being made. As well as keep track of any design decisions being made as work is iterated on.

DennisDenuto committed:
Feat: Allow setting registry auth via env variables

- brought in env keychain from kbld
- backfilled env keychain tests
- use a MultiKeychain used by the registry.go

Co-authored-by: Dennis Leon <leonde@vmware.com>
Co-authored-by: Steven Locke <slocke@pivotal.io>

DennisDenuto committed:
Backfill test

- Added test for Registry and identity token used in env keychain
- Add e2e test for Basic, and Bearer token auth flows
- Add basic auth capabilities to the fake registry and use that in our e2e test
- Add e2e test that tests using the IMGPKG_REGISTRY_* env variables to auth

- Test refactor
-- use a constant as the environment key name prefix
-- use a function to unset test environment variables used

Co-authored-by: Dennis Leon <leonde@vmware.com>
Co-authored-by: Steven Locke <slocke@pivotal.io>

DennisDenuto commented:
TODO:

- [x] Rebase commits
- [x] Do the TODO's in the codebase commits 
- [x] Update documentation

DennisDenuto committed:
UX change

-- change the order of precedence when both the auth flags and auth env varaibles are set.
- Test refactor
-- Make test name consistent (stop using snakecase)

Co-authored-by: Dennis Leon <leonde@vmware.com>
Co-authored-by: Steven Locke <slocke@pivotal.io>

DennisDenuto committed:
Code cleanup

- Fix typos
- remove stale TODO

Co-authored-by: Dennis Leon <leonde@vmware.com>
Co-authored-by: Steven Locke <slocke@pivotal.io>

DennisDenuto committed:
Code Refactor

- custom keychain and env keychain are now abstracted by an
image.Keychain function. This encapsulates the ordering of which
'keychain' is used when both flags and env variables are provided to
imgpkg
- This also made testing our mult-keychain logic that our registry
depends on much easier
- move registry and keychain into their own 'registry' package
- pass list of environment variables to the keychain (instead of calling
os.Environ itself)
- use string pre/suffix to get the authOpt instead of looping over the funcsMap

DennisDenuto commented:
Created a [PR](https://github.com/vmware-tanzu/carvel.dev/pull/85) for doc changes

cppforlife commented on pkg/imgpkg/image/keychain.go:
```diff
@@ -0,0 +1,208 @@
+	globalPrefix = "IMGPKG_REGISTRY"
+)
+
+func Keychain(username string, password string, token string, anon bool) regauthn.Keychain {
```
- it might be good to take a struct with named fields since there are 3 fields that have the same type (could be opts struct that customRegistryKeychain uses). i typically recommend this when types by themselves arent distinguishing enough.
- might be good to take a `environFunc func() []string` argument that would be set to os.Environ and some fake function in tests so that tests do not have to fiddle with real environment of the process (aka global state).

joaopapereira replied:
I like the second suggestion because this function can already filter out the environment variables that we care about so that we do not have to go through all of them.

cppforlife replied:
imho filtering still belongs in this class since it's related to interpreting various env vars names. i was trying to highlight that we can do dependency injection (for reading env vars) via a function arg.

cppforlife commented on pkg/imgpkg/image/keychain.go:
```diff
@@ -0,0 +1,208 @@
+	}
+}
+
+func retryDefaultKeychain(doFunc func() (regauthn.Authenticator, error)) (regauthn.Authenticator, error) {
```
lets put retryDefaultKeychain on customRegistryKeychain since it's only used by this class to indicate that its only used there.

cppforlife commented on pkg/imgpkg/image/keychain.go:
```diff
@@ -0,0 +1,208 @@
+	// ErrCredentialsMissingServerURL and ErrCredentialsMissingUsername standardize
+	// invalid credentials or credentials management operations
+	errCredentialsMissingServerURLMessage = "no credentials server URL"
+	errCredentialsMissingUsernameMessage  = "no credentials username"
```
we can probably localize this as consts in the retry function since that's where it's only being used.

cppforlife commented on pkg/imgpkg/image/keychain.go:
```diff
@@ -0,0 +1,208 @@
+	errCredentialsMissingServerURLMessage = "no credentials server URL"
+	errCredentialsMissingUsernameMessage  = "no credentials username"
+
+	globalPrefix = "IMGPKG_REGISTRY"
```
i think this can also be inlined since it's only used in one place.

cppforlife commented on pkg/imgpkg/image/keychain.go:
```diff
@@ -0,0 +1,208 @@
+// Copyright 2020 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+package image
```
it might be a good time to move registry and keychain stuff into their own package since they are both not that related to other classes in image package. i can even see that package being shared between imgpkg and kbld.

cppforlife commented on pkg/imgpkg/image/keychain_test.go:
```diff
@@ -0,0 +1,136 @@
+// Copyright 2021 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+package image
```
should be in its own package image_test to avoid accidental private state use from image package.

joaopapereira commented on pkg/imgpkg/image/keychain.go:
```diff
@@ -0,0 +1,208 @@
+
+	if k.collected || k.collectErr != nil {
+		return append([]envKeychainInfo{}, k.infos...), k.collectErr
+	}
```
```suggestion
	if k.collected {
		return append([]envKeychainInfo{}, k.infos...), nil
	}

	if k.collectErr != nil {
		return nil, k.collectErr
	}
```
I am suggesting this change because it is clearer, at least in my head these 2 possible states. If not you will have to look through the code to see if there is a case where an error has occurred and we collected the keychain information.

joaopapereira commented on pkg/imgpkg/image/keychain_test.go:
```diff
@@ -0,0 +1,136 @@
+}
+
+func setTestEnv(t *testing.T, key string, value string) {
+	assert.NoError(t, os.Setenv(globalPrefix+"_"+key, value))
```
```suggestion
	assert.NoError(t, os.Setenv("IMGPKG_REGISTRY_"+key, value))
```

For testing purposes, I prefer to have hardcoded the constant strings. This allows us to catch errors if for some reason these constants change or are incorrect.

joaopapereira commented on pkg/imgpkg/image/keychain_test.go:
```diff
@@ -0,0 +1,136 @@
+
+func unsetTestEnv(t *testing.T) {
+	for _, env := range os.Environ() {
+		if strings.HasPrefix(env, globalPrefix) {
```
```suggestion
		if strings.HasPrefix(env, "IMGPKG_REGISTRY_") {
```
For testing purposes, I prefer to have hardcoded the constant strings. This allows us to catch errors if for some reason these constants change or are incorrect.

joaopapereira commented on pkg/imgpkg/image/keychain.go:
```diff
@@ -0,0 +1,208 @@
+
+		var matched bool
+
+		for key, updateFunc := range funcsMap {
```
Since `funcsMap` is a Map I think that there should be a more efficient way to do this

joaopapereira reviewed (changes requested):
I agree with @cppforlife I think we should move the keychain and the repository from the image package since it feels out of place in that package.
Also left some change suggestions

StevenLocke commented on pkg/imgpkg/registry/keychain.go:
```diff
@@ -0,0 +1,217 @@
+					return nil, k.collectErr
+				}
+			} else {
+				splits := strings.SplitN(authOpt, "_", 2)
```
This isn't quite right. It would work if our "types" didn't include `_` themselves, but some do. Whoops! Back to the drawing board.

cppforlife commented:
@joaopapereira @StevenLocke my 2c: this (parsing env vars) is not perf sensitive code so readability here should be prime objective.

joaopapereira commented:
> @joaopapereira @StevenLocke my 2c: this (parsing env vars) is not perf sensitive code so readability here should be prime objective.

After looking at the implementation there was a point that I overlooked. So maybe we can revert back to the code we had there and when we do that I think we are ok to merge this PR

joaopapereira commented on pkg/imgpkg/registry/keychain.go:
```diff
@@ -0,0 +1,214 @@
+	return append([]envKeychainInfo{}, k.infos...), nil
+}
+
+type envKeychainInfo struct {
```
Lets move this definition near the place it is used, the prior struct

DennisDenuto replied:
Done

cppforlife commented on pkg/imgpkg/imagetar/tar_writer.go:
```diff
@@ -202,22 +202,22 @@ func (w *TarWriter) writeLayers() error {
 
 func (w *TarWriter) fillInLayers(writtenLayers map[string]writtenLayer) error {
-	var sortedWritterLayers []writtenLayer
+	var sortedWriterLayers []writtenLayer
```
this was meant to be `sortedWrittenLayers`

cppforlife commented on pkg/imgpkg/registry/keychain.go:
```diff
@@ -0,0 +1,214 @@
+}
+
+func Keychain(keychainOpts KeychainOpts, environFunc func() []string) regauthn.Keychain {
+	return regauthn.NewMultiKeychain(customRegistryKeychain{opts: keychainOpts},
```
it seems that customRegistryKeychain falls back to retrying default keychain. this would happen before trying to access envKeychain based on my understanding of NewMultiKeychain. are the tests falling to envKeychain because regauthn.DefaultKeychain is set to anon? this seems like a pretty nasty global variable.

joaopapereira replied:
Also if you set Anon to true, it will try the next keychain in the list

DennisDenuto replied:
Ok i'll add some logic around handling cases when different overlapping credentials are provided to imgpkg. Namely, cli flags, env variables and config.json.

I'll work towards meeting the following cases.

## case 1) username + password is provided via cli flags
MultiKeychain uses customKeyChain

## case 2) anon is provided via cli flags
An "Anonymous" keychain is used (notice that MultiKeychain isn't used)

## case 3) env credentials are provided with the anon cli flag
e.g.
```
REGISTRY_HOSTNAME=some.reg
REGISTRY_USERNAME=foo
imgpkg pull --anon
```
An "Anonymous" keychain is used (notice that MultiKeychain isn't used)

## case 4) env credentials are provided, *and* credentials are also contained in `config.json`
MultiKeychain traverses the keychain (custom keychain -> env keychain -> default keychain) until it finds a keychain to resolve the resource




Note: An "Anonymous keychain" is a yet-to-be created keychain that resolves every resource to "Anonymous"

cppforlife replied:
> An "Anonymous" keychain is used (notice that MultiKeychain isn't used)

im not sure we would want that. e.g. you might be configuring anonymous access for everything except for providing credentials for a specific registry. it seems at high level that MultiKeyChain does not capture "i found anon case and i want to use it" and instead treats anon case as fallthrough to the next option. how can we continue using MultiKeyChain and support our "explicit" anon case?

DennisDenuto replied:
Ok.

I'll make the ordering the following:

```
1. registry matching feature
  a. IMGPKG_USERNAME_0
2. generic auth
  a. IMGPKG_USERNAME
  b. --username <-- would take precedence over (a)
3. anon flag
4. default
```

joaopapereira commented on pkg/imgpkg/registry/keychain_test.go:
```diff
@@ -0,0 +1,121 @@
+	assert.Equal(t, authn.Anonymous, auth)
+}
+
+func TestEnvAuthWhenEnvVarsProvided(t *testing.T) {
```
Can we add in one of these tests an environment variable that looks like "SOMETHING_USERNAME" to ensure we are only considering the ones that start with `IMGPKG_REGISTRY`?

DennisDenuto replied:
Ok done

joaopapereira commented on pkg/imgpkg/registry/keychain_test.go:
```diff
@@ -0,0 +1,121 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func TestAnonAuthWhenNoEnvVarsProvided(t *testing.T) {
```
Please refactor these tests to use the `t.Run` format that we have been using in `imgpkg`

DennisDenuto replied:
Good idea. Also added a parent test context to test the rules around ordering around using multiple keychains

joaopapereira commented on test/e2e/auth_test.go:
```diff
@@ -0,0 +1,77 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func TestBasicAuth(t *testing.T) {
```
I would change these tests to also follow the `t.Run` layout we are using in other tests.

DennisDenuto replied:
done.

joaopapereira reviewed (changes requested):
Please consolidate the commits, so that we do not have the extra commit for the change of the map, and the subsequent rollback commit.

DennisDenuto committed:
Feat: Specify the ordering of authentication

- When there are multiple auth's provided the following is the order that
will be followed to determine which credentials to use.

1. registry matching feature
  a. IMGPKG_USERNAME_0
2. generic auth
  a. IMGPKG_USERNAME
  b. --username <-- would take precedence over (a)
3. anon flag
4. config.json

- Code refactor
-- test refactor. Tests have a parent context and uses t.Run
-- Ordered definitions of different keychains to match the order they are
used

Authored-by: Dennis Leon <leonde@vmware.com>

joaopapereira reviewed (approved):
LGTM
